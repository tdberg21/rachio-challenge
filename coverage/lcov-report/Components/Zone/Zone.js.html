<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for Components/Zone/Zone.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">Components/Zone</a> Zone.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">93.75% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>15/16</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">87.5% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>7/8</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">93.75% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>15/16</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">11x</span>
<span class="cline-any cline-yes">11x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { Component } from 'react';
import { apiKey } from '../../helpers/apiKey.js';
import './Zone.css';
import PropTypes from 'prop-types';
&nbsp;
export default class Zone extends Component {
  constructor() {
    super();
&nbsp;
    this.state = {
      message: '',
      duration: 0
    };
&nbsp;
    this.handleActivateZone = this.handleActivateZone.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }
&nbsp;
  handleChange(event) {
    let duration = event.target.value;
    this.setState({
      duration
    });
  }
&nbsp;
  handleActivateZone(event) {
    const id = event.target.value;
    const duration = parseInt(this.state.duration);
    fetch('https://api.rach.io/1/public/zone/start', {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`
      },
      body: JSON.stringify({
        id,
        duration
      })
    })
      .then(() =&gt; this.setState({message: 'Success!'}))
      .catch(<span class="fstat-no" title="function not covered" >er</span>ror =&gt; <span class="cstat-no" title="statement not covered" >this.setState({message: `Error: ${error}`}))</span>;
  }
&nbsp;
  render() {
    let { zone } = this.props;
    return (
      &lt;div&gt;
        &lt;h4&gt;{zone.name}&lt;/h4&gt;
        &lt;img src={zone.imageUrl} alt={zone.name} className='zone-image'/&gt;
        &lt;p&gt;Max Runtime: {zone.maxRuntime} &lt;/p&gt;
        &lt;label htmlFor='duration'&gt;Runtime Duration:&lt;/label&gt;
        &lt;input
          type='number'
          id='duration'
          className='duration-input'
          placeholder='Desired Runtime'
          value={this.state.duration}
          name='duration'
          onChange={(event) =&gt; this.handleChange(event)}&gt;
        &lt;/input&gt;
        &lt;button 
          onClick={(event) =&gt; this.handleActivateZone(event)} 
          className='activate-zone-button'
          value={zone.id}&gt;
          Activate Zone
        &lt;/button&gt;
        {this.state.message}
      &lt;/div&gt;
    );
  }
}
&nbsp;
Zone.propTypes = {
  // zone: PropTypes.obj,
  duration: PropTypes.number
};</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Oct 04 2018 13:07:58 GMT-0600 (MDT)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
