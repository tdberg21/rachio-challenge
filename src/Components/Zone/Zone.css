.zone-image {
  height: 150px;
}

.zone-component {
  margin-top: 10px;
  border-top: 2px solid #767679;
  display: flex;
}

.zone-sections {
  width: 50%;
  padding-bottom: 15px;
}

.zone-info-section {
  padding-left: 25px;
}

.zone-form-section {
  display: flex;
  flex-direction: column;
  padding-top: 30px;
}