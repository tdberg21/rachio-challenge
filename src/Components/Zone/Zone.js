import React, { Component } from 'react';
import { apiKey } from '../../helpers/apiKey.js';
import './Zone.css';
import PropTypes from 'prop-types';

export default class Zone extends Component {
  constructor() {
    super();

    this.state = {
      message: '',
      duration: 0
    };

    this.handleActivateZone = this.handleActivateZone.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    let duration = event.target.value;
    this.setState({
      duration
    });
  }

  handleActivateZone(event) {
    const id = event.target.value;
    const duration = parseInt(this.state.duration);
    fetch('https://api.rach.io/1/public/zone/start', {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`
      },
      body: JSON.stringify({
        id,
        duration
      })
    })
      .then(() => this.setState({message: 'Success!'}))
      .catch(error => this.setState({message: `Error: ${error}`}));
  }

  render() {
    let { zone } = this.props;
    return (
      <div>
        <h4>{zone.name}</h4>
        <img src={zone.imageUrl} alt={zone.name} className='zone-image'/>
        <p>Max Runtime: {zone.maxRuntime} </p>
        <label htmlFor='duration'>Runtime Duration:</label>
        <input
          type='number'
          id='duration'
          className='duration-input'
          placeholder='Desired Runtime'
          value={this.state.duration}
          name='duration'
          onChange={(event) => this.handleChange(event)}>
        </input>
        <button 
          onClick={(event) => this.handleActivateZone(event)} 
          className='activate-zone-button'
          value={zone.id}>
          Activate Zone
        </button>
        {this.state.message}
      </div>
    );
  }
}

Zone.propTypes = {
  // zone: PropTypes.obj,
  duration: PropTypes.number
};