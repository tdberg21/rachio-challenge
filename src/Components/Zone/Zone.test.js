import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Zone from './Zone.js';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

describe('Zone tests', () => {
  let wrapper;
  let mockZone = {
    enabled: false,
    id: "bf329221",
    imageUrl: "google.com",
    lastWateredDate: 1535642590000,
    maxRuntime: 10800,
    name: "Zone 2",
    zoneNumber: 2
  };

  beforeEach(() => {
    wrapper = shallow(<Zone zone={mockZone} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when handleChange is called', () => {
    const mockEvent = { target: { value: 60 } };
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('duration')).toBe(60);
  });

  it('should invoke handleChange when runtime duration is changed', () => {
    const spy = spyOn(wrapper.instance(), 'handleChange');
    wrapper.instance().forceUpdate();
    const mockEvent = { target: { value: 'something' } };

    wrapper.find('.duration-input').simulate('change', mockEvent);

    expect(spy).toHaveBeenCalled();
  });

  it('should make a fetch request when handleActivateZone is called', () => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({})
    }));
    const mockEvent = { target: {value: 2} };
    wrapper.instance().handleActivateZone(mockEvent);

    expect(window.fetch).toHaveBeenCalled();
  });

  it('should update state with a message of success when handleActivateZone is called and there is no error', () => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({})
    }));
    const mockEvent = { target: { value: 2 } };
    
    window.fetch = jest.fn().mockImplementationOnce(() => Promise.resolve({}));
    Promise.resolve(wrapper.instance().handleActivateZone(mockEvent))
      .then(() => {
        wrapper.update();
      })
      .then(() => {
        wrapper.update();
      })
      .then(() => {
        expect(wrapper.state('message')).toEqual('Success!');
      });
  });

  it('should update state with a message of success when handleActivateZone is called and there is an error', () => {
    const mockEvent = { target: { value: 2 } };
    window.fetch = jest.fn().mockImplementationOnce(() => Promise.reject(
      new Error('failed')
    ));
    Promise.resolve(wrapper.instance().handleActivateZone(mockEvent))
      .then(() => {
        wrapper.update();
      })
      .then(() => {
        wrapper.update();
      })
      .then(() => {
        expect(wrapper.state('message')).toEqual('Error!');
      });
  });

  it('should invoke handleActivateZone when activate zone button is clicked', () => {
    const spy = spyOn(wrapper.instance(), 'handleActivateZone');
    const mockEvent = { target: { value: 2 } };
    wrapper.instance().forceUpdate();

    wrapper.find('.activate-zone-button').simulate('click', mockEvent);

    expect(spy).toHaveBeenCalled();
  });
});

