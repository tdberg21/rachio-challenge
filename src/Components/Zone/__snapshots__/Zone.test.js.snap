// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zone tests should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Zone
    zone={
      Object {
        "enabled": false,
        "id": "bf329221",
        "imageUrl": "google.com",
        "lastWateredDate": 1535642590000,
        "maxRuntime": 10800,
        "name": "Zone 2",
        "zoneNumber": 2,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h4>
          Zone 2
        </h4>,
        <img
          alt="Zone 2"
          className="zone-image"
          src="google.com"
        />,
        <p>
          Max Runtime: 
          10800
           
        </p>,
        <label
          htmlFor="duration"
        >
          Runtime Duration:
        </label>,
        <input
          className="duration-input"
          id="duration"
          name="duration"
          onChange={[Function]}
          placeholder="Desired Runtime"
          type="number"
          value={0}
        />,
        <button
          className="activate-zone-button"
          onClick={[Function]}
          value="bf329221"
        >
          Activate Zone
        </button>,
        "",
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Zone 2",
        },
        "ref": null,
        "rendered": "Zone 2",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "Zone 2",
          "className": "zone-image",
          "src": "google.com",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Max Runtime: ",
            10800,
            " ",
          ],
        },
        "ref": null,
        "rendered": Array [
          "Max Runtime: ",
          10800,
          " ",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Runtime Duration:",
          "htmlFor": "duration",
        },
        "ref": null,
        "rendered": "Runtime Duration:",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "duration-input",
          "id": "duration",
          "name": "duration",
          "onChange": [Function],
          "placeholder": "Desired Runtime",
          "type": "number",
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Activate Zone",
          "className": "activate-zone-button",
          "onClick": [Function],
          "value": "bf329221",
        },
        "ref": null,
        "rendered": "Activate Zone",
        "type": "button",
      },
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4>
            Zone 2
          </h4>,
          <img
            alt="Zone 2"
            className="zone-image"
            src="google.com"
          />,
          <p>
            Max Runtime: 
            10800
             
          </p>,
          <label
            htmlFor="duration"
          >
            Runtime Duration:
          </label>,
          <input
            className="duration-input"
            id="duration"
            name="duration"
            onChange={[Function]}
            placeholder="Desired Runtime"
            type="number"
            value={0}
          />,
          <button
            className="activate-zone-button"
            onClick={[Function]}
            value="bf329221"
          >
            Activate Zone
          </button>,
          "",
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Zone 2",
          },
          "ref": null,
          "rendered": "Zone 2",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "Zone 2",
            "className": "zone-image",
            "src": "google.com",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Max Runtime: ",
              10800,
              " ",
            ],
          },
          "ref": null,
          "rendered": Array [
            "Max Runtime: ",
            10800,
            " ",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Runtime Duration:",
            "htmlFor": "duration",
          },
          "ref": null,
          "rendered": "Runtime Duration:",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "duration-input",
            "id": "duration",
            "name": "duration",
            "onChange": [Function],
            "placeholder": "Desired Runtime",
            "type": "number",
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Activate Zone",
            "className": "activate-zone-button",
            "onClick": [Function],
            "value": "bf329221",
          },
          "ref": null,
          "rendered": "Activate Zone",
          "type": "button",
        },
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
